modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -3135,12 +3156,24 @@ def groupby_agg_builder(df, by=None, drop=False, partition_idx=None):
                 by = []
 
             by += not_broadcastable_by
+            if level is not None:
+                assert not by, "When handling simple level groupby, `by` must be empty"
+                by = df.axes[axis].get_level_values(level)
```

Comment:
I'm not exactly sure how this could have worked, as `pandas.groupby()` now bails if `level` is non-`None` and `by` is non-`None` (even if `by` is an empty list).

Also, while technically we could pass the `level` argument instead of getting the list of values (and handle it like _not_ pass the `by` if `level` is non-`None`), the code downwards relies on `by` having a valid length.

Reply:
I've been able to make this work when I explicitly replaced an empty list with `None` in case of a `level` was passed:
<details><summary>git diff</summary>

```diff
diff --git a/modin/core/storage_formats/pandas/query_compiler.py b/modin/core/storage_formats/pandas/query_compiler.py
index 80b4c7b6..8188ea32 100644
--- a/modin/core/storage_formats/pandas/query_compiler.py
+++ b/modin/core/storage_formats/pandas/query_compiler.py
@@ -3085,8 +3085,6 @@ class PandasQueryCompiler(BaseQueryCompiler):
         groupby_kwargs = groupby_kwargs.copy()

         as_index = groupby_kwargs.get("as_index", True)
-        # remove 'level' from kwargs as we're going to handle it separately
-        groupby_kwargs.pop("level", None)
         by, internal_by = self._groupby_internal_columns(by, drop)

         broadcastable_by = [o._modin_frame for o in by if isinstance(o, type(self))]
@@ -3156,9 +3154,8 @@ class PandasQueryCompiler(BaseQueryCompiler):
                 by = []

             by += not_broadcastable_by
-            if level is not None:
-                assert not by, "When handling simple level groupby, `by` must be empty"
-                by = df.axes[axis].get_level_values(level)
+            if level is not None and not by:
+                by = None

             def compute_groupby(df, drop=False, partition_idx=0):
                 """Compute groupby aggregation for a single partition."""
@@ -3205,7 +3202,7 @@ class PandasQueryCompiler(BaseQueryCompiler):
                         result,
                         internal_by_cols,
                         by_cols_dtypes=df[internal_by_cols].dtypes.values,
-                        by_length=len(by),
+                        by_length=len(by) if by is not None else 0,
                         selection=selection,
                         partition_idx=partition_idx,
                         drop=drop,
```

</details>

My main complaint regarding your approach is that getting an index level is not free as MultiIndex levels are usually dictionary encoded, meaning that `.get_level_values` requires materializing an actual index array:
```python
import pandas as pd
import numpy as np
import timeit

N = 10_000_000
df = pd.DataFrame({"a": np.arange(N)}, index=pd.MultiIndex.from_product([["a", "b"], np.arange(N // 2)]))
print(timeit.timeit(lambda: df.index.get_level_values(0), number=10)) # ~1.01s
```

Reply:
This was my original intention, though I thought `by` is going to be used in more places 😄 

That said, when I remove defaulting to pandas on non-empty `level` at the top (what you didn't do in your diff BTW), one test fails for me (so I still have to fix that):
`pytest --execution=PandasOnPython "modin/pandas/test/test_groupby.py::test_mixed_dtypes_groupby[False]" --pdb`

This is being caused by the logic not resetting the index in this case, as we need valid (non-zero) `by_length` for which I do not see any valid way of computing the length aside from getting values using `.get_level_values()`...

This is my latest patch:
```diff
diff --git a/modin/core/storage_formats/pandas/query_compiler.py b/modin/core/storage_formats/pandas/query_compiler.py
index 8188ea32..775fa5fe 100644
--- a/modin/core/storage_formats/pandas/query_compiler.py
+++ b/modin/core/storage_formats/pandas/query_compiler.py
@@ -3042,22 +3042,11 @@ class PandasQueryCompiler(BaseQueryCompiler):
         drop=False,
         series_groupby=False,
     ):
-        level = groupby_kwargs.get("level", None)
-        if is_list_like(level) and len(level) == 1:
-            level = level[0]
-        # We default to pandas in complex cases of level groupby, i.e. when level is given and
-        # it is not an integer in a multi-index axis or by is not None.
-        complex_level = level is not None and (
-            by is not None
-            or not isinstance(level, int)
-            or not self.has_multiindex(axis)
-        )
-
-        # Also defaulting to pandas in case of an empty frame as we can't process it properly.
+        # Defaulting to pandas in case of an empty frame as we can't process it properly.
         # Higher API level won't pass empty data here unless the frame has delayed
         # computations. So we apparently lose some laziness here (due to index access)
         # because of the inability to process empty groupby natively.
-        if complex_level or len(self.columns) == 0 or len(self.index) == 0:
+        if len(self.columns) == 0 or len(self.index) == 0:
             return super().groupby_agg(
                 by, agg_func, axis, groupby_kwargs, agg_args, agg_kwargs, how, drop
             )
@@ -3154,7 +3143,7 @@ class PandasQueryCompiler(BaseQueryCompiler):
                 by = []

             by += not_broadcastable_by
-            if level is not None and not by:
+            if groupby_kwargs.get("level", None) is not None and not by:
                 by = None

             def compute_groupby(df, drop=False, partition_idx=0):
@@ -3202,7 +3191,7 @@ class PandasQueryCompiler(BaseQueryCompiler):
                         result,
                         internal_by_cols,
                         by_cols_dtypes=df[internal_by_cols].dtypes.values,
-                        by_length=len(by) if by is not None else 0,
+                        by_length=len(by) if by is not None else None,
                         selection=selection,
                         partition_idx=partition_idx,
                         drop=drop,
```

Reply:
So the way to compute valid `by_length` for using level parameter would be:
```python
level = groupby_kwargs.get("level", None)
if level is not None and not by:
    by = None
    by_length = len(level) if is_list_like(level) else 1
else:
    by_length = len(by)
```
Note that `by_length` has to indicate the number of "columns" you're grouping on, not the length of the grouper itself. 

You would additionally need to add a check for whether the kernel you're running is a transform-like aggregation, as pandas never reset the index of the transform results (I don't know how the tests used to work without this `is_transform` check, probably they were just defaulting to pandas).

So summarizing this, here are the final diffs that include all the mentioned changes. I've run groupby and rolling tests for them and they all have passed.

<details><summary>git diff</summary>

```diff
diff --git a/modin/core/dataframe/algebra/default2pandas/groupby.py b/modin/core/dataframe/algebra/default2pandas/groupby.py
index d91a668b..b5ae3a1e 100644
--- a/modin/core/dataframe/algebra/default2pandas/groupby.py
+++ b/modin/core/dataframe/algebra/default2pandas/groupby.py
@@ -433,7 +433,7 @@ class GroupBy:
         if by_length is None:
             by_length = len(internal_by_cols)

-        reset_index = by_length > 0 or selection is not None
+        reset_index = method != "transform" and (by_length > 0 or selection is not None)

         # If the method is "size" then the result contains only one unique named column
         # and we don't have to worry about any naming conflicts, so inserting all of
diff --git a/modin/core/storage_formats/pandas/query_compiler.py b/modin/core/storage_formats/pandas/query_compiler.py
index 80b4c7b6..839c6210 100644
--- a/modin/core/storage_formats/pandas/query_compiler.py
+++ b/modin/core/storage_formats/pandas/query_compiler.py
@@ -39,6 +39,7 @@ from collections.abc import Iterable
 from typing import List, Hashable
 import warnings
 import hashlib
+from pandas.core.groupby.base import transformation_kernels

 from modin.core.storage_formats.base.query_compiler import BaseQueryCompiler
 from modin.config import Engine
@@ -3042,22 +3043,11 @@ class PandasQueryCompiler(BaseQueryCompiler):
         drop=False,
         series_groupby=False,
     ):
-        level = groupby_kwargs.get("level", None)
-        if is_list_like(level) and len(level) == 1:
-            level = level[0]
-        # We default to pandas in complex cases of level groupby, i.e. when level is given and
-        # it is not an integer in a multi-index axis or by is not None.
-        complex_level = level is not None and (
-            by is not None
-            or not isinstance(level, int)
-            or not self.has_multiindex(axis)
-        )
-
         # Also defaulting to pandas in case of an empty frame as we can't process it properly.
         # Higher API level won't pass empty data here unless the frame has delayed
         # computations. So we apparently lose some laziness here (due to index access)
         # because of the inability to process empty groupby natively.
-        if complex_level or len(self.columns) == 0 or len(self.index) == 0:
+        if len(self.columns) == 0 or len(self.index) == 0:
             return super().groupby_agg(
                 by, agg_func, axis, groupby_kwargs, agg_args, agg_kwargs, how, drop
             )
@@ -3067,6 +3057,8 @@ class PandasQueryCompiler(BaseQueryCompiler):
                 by, agg_func, axis, groupby_kwargs, agg_args, agg_kwargs, drop
             )

+        is_transform_method = how == "transform" or (isinstance(agg_func, str) and agg_func in transformation_kernels)
+
         if isinstance(agg_func, dict):
             assert (
                 how == "axis_wise"
@@ -3085,8 +3077,6 @@ class PandasQueryCompiler(BaseQueryCompiler):
         groupby_kwargs = groupby_kwargs.copy()

         as_index = groupby_kwargs.get("as_index", True)
-        # remove 'level' from kwargs as we're going to handle it separately
-        groupby_kwargs.pop("level", None)
         by, internal_by = self._groupby_internal_columns(by, drop)

         broadcastable_by = [o._modin_frame for o in by if isinstance(o, type(self))]
@@ -3156,9 +3146,12 @@ class PandasQueryCompiler(BaseQueryCompiler):
                 by = []

             by += not_broadcastable_by
-            if level is not None:
-                assert not by, "When handling simple level groupby, `by` must be empty"
-                by = df.axes[axis].get_level_values(level)
+            level = groupby_kwargs.get("level", None)
+            if level is not None and not by:
+                by = None
+                by_length = len(level) if is_list_like(level) else 1
+            else:
+                by_length = len(by)

             def compute_groupby(df, drop=False, partition_idx=0):
                 """Compute groupby aggregation for a single partition."""
@@ -3205,11 +3198,12 @@ class PandasQueryCompiler(BaseQueryCompiler):
                         result,
                         internal_by_cols,
                         by_cols_dtypes=df[internal_by_cols].dtypes.values,
-                        by_length=len(by),
+                        by_length=by_length,
                         selection=selection,
                         partition_idx=partition_idx,
                         drop=drop,
                         inplace=True,
+                        method="transform" if is_transform_method else None
                     )
                 else:
                     new_index_names = tuple(
```

</details>

Reply:
Thanks @dchigarev, this helped a lot!